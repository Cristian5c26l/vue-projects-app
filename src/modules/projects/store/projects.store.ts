import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import { v4 as uuidv4 } from 'uuid';
import { useLocalStorage } from '@vueuse/core';
import type { Project } from '../interfaces/project.interface';

// const initialLoad = (): Project[] => {
//   return [
//     {
//       id: uuidv4(),
//       name: 'Project 1',
//       tasks: [],
//     },
//     {
//       id: uuidv4(),
//       name: 'Project 2',
//       tasks: [],
//     },
//     // {
//     //   id: '3',
//     //   name: 'Project 3',
//     //   tasks: [],
//     // },
//   ];
// };

export const useProjectsStore = defineStore('projects', () => {
  // const projects = ref(useLocalStorage<Project[]>('projects', initialLoad())); // Se espera que el item "projects" de local storage sea un array de proyectos. En caso de que no exista, se inicializa con el array de proyectos retornado por la función initialLoad. Cabe decir que, con solo actualizar la variable reactiva "projects", este projects.value se encargará de ser el nuevo valor del item "projects" en local storage.
  const projects = ref(useLocalStorage<Project[]>('projects', []));

  const addProject = (name: string) => {
    if (name.length === 0) return;
    projects.value.push({
      id: uuidv4(),
      name,
      tasks: [],
    });
  };

  const addTaskToProject = (projectId: string, taskName: string) => {
    if (taskName.trim().length === 0) return;

    const project = projects.value.find((p) => p.id === projectId);

    if (!project) return;

    project.tasks.push({
      id: uuidv4(),
      name: taskName,
    });
  };

  const toggleTask = (projectId: string, taskId: string) => {
    const project = projects.value.find((p) => p.id === projectId);
    if (!project) return;

    const task = project.tasks.find((t) => t.id === taskId);
    if (!task) return;

    task.completedAt = task.completedAt ? undefined : new Date();
  };

  return {
    // Properties
    projects,

    // Getters
    projectList: computed(() => [...projects.value]),
    noProjects: computed(() => projects.value.length === 0),
    // projectsWithCompletion: computed(() => {
    // {
    //   id,
    //   name,
    //   taskCount,
    //   completion,
    // }
    // }),
    // Numero de Tareas del proyecto -> 100 %
    // Tareas Completadas -> x
    // x = Tareas Completadas * 100 / Numero de Tareas del proyecto
    // projectsWithCompletion: computed(() =>
    //   [...projects.value].map((project) => ({
    //     id: project.id,
    //     name: project.name,
    //     taskCount: project.tasks.length,
    //     completion:
    //       project.tasks.length === 0
    //         ? 0
    //         : (project.tasks.filter((tasks) => !!tasks.completedAt).length * 100) /
    //           project.tasks.length,
    //   })),
    // ),
    projectsWithCompletion: computed(() => {
      return projects.value.map((project) => {
        const taskCount = project.tasks.length;
        const completedTasks = project.tasks.filter((t) => t.completedAt).length;
        const completion = taskCount === 0 ? 0 : (completedTasks / taskCount) * 100;

        return {
          id: project.id,
          name: project.name,
          taskCount,
          completion,
        };
      });
    }),

    // Actions (Methods)
    addProject: addProject,
    addTaskToProject,
    toggleTask,
  };
});
