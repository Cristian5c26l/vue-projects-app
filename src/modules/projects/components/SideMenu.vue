<template>
  <aside class="bg-base-200 w-72 min-h-screen">
    <h2 class="text-lg font-bold mx-4"><RouterLink to="/">Proyectos</RouterLink></h2>
    <p v-if="projectsStore.noProjects" class="text-sm text-gray-500 mx-4">No hay proyectos</p>

    <!-- Menu -->
    <ul v-else class="menu">
      <!-- Cada li con template (que solo agrupa a details), es un menú desplegable de un proyecto con sus tareas-->
      <li v-for="project in projectsStore.projects" :key="project.id">
        <!--template aquí sirve como agrupador sin representacion HTML. Solo sirve para agrupar elementos (el div agrupa elementos y tiene representacion HTML).-->
        <template v-if="project.tasks.length > 0">
          <details>
            <summary>
              <RouterLink :to="`/project/${project.id}`">{{ project.name }}</RouterLink>
            </summary>
            <ul>
              <li v-for="task in project.tasks" :key="task.id">
                <RouterLink :to="`/project/${project.id}`">{{ task.name }}</RouterLink>
              </li>
            </ul>
          </details>
        </template>
        <!--Enlace al proyecto solamente, en caso de que no tenga tareas-->
        <template v-else>
          <RouterLink :to="`/project/${project.id}`">{{ project.name }}</RouterLink>
        </template>
      </li>
    </ul>
  </aside>
</template>

<script lang="ts" setup>
import { useProjectsStore } from '../store/projects.store';

const projectsStore = useProjectsStore();
</script>
