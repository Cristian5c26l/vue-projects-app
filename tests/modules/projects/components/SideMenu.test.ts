import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import SideMenu from '@/modules/projects/components/SideMenu.vue';
import { useProjectsStore } from '@/modules/projects/store/projects.store';
import { fakeProjects } from '../../../mocks/projects.fake';
import { nextTick } from 'vue';

describe('<SideMenu />', () => {
  const wrapper = mount(SideMenu, {
    global: {
      stubs: ['RouterLink'],
      plugins: [createTestingPinia()],
    },
  }); // createTestingPinia crea la pinia, la cual va a contener en este caso, el store "projectsStore".

  const projectsStore = useProjectsStore();

  beforeEach(() => {
    projectsStore.$patch({
      projects: [],
    });
  });

  test('should render with no projects', () => {
    // console.log(wrapper.html());

    expect(wrapper.html()).toContain('No hay proyectos');
  });

  test('should render with projects', async () => {
    projectsStore.$patch({
      projects: fakeProjects,
    }); // Cambio en el state de projectsStore es s√≠ncrono, pero en lo que respecta al html es asincrono, por lo que hay que usar nextTick con await para esperar a que el cambio del state se vea reflejado en el html.

    // console.log(projectsStore.projectList);

    await nextTick();

    // console.log(wrapper.html());
    expect(wrapper.html()).not.toContain('No hay proyectos');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
