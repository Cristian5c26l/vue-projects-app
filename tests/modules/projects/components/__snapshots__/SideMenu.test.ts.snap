// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SideMenu /> > should render with projects 1`] = `
"<aside class="bg-base-200 w-72 min-h-screen">
  <h2 class="text-lg font-bold mx-4">
    <router-link-stub to="/"></router-link-stub>
  </h2><!-- Menu -->
  <ul class="menu">
    <!-- Cada li con template (que solo agrupa a details), es un menú desplegable de un proyecto con sus tareas-->
    <li>
      <!--template aquí sirve como agrupador sin representacion HTML. Solo sirve para agrupar elementos (el div agrupa elementos y tiene representacion HTML).-->
      <details>
        <summary>
          <router-link-stub to="/project/1"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
    <li>
      <!--template aquí sirve como agrupador sin representacion HTML. Solo sirve para agrupar elementos (el div agrupa elementos y tiene representacion HTML).-->
      <!--Enlace al proyecto solamente, en caso de que no tenga tareas-->
      <router-link-stub to="/project/2"></router-link-stub>
    </li>
    <li>
      <!--template aquí sirve como agrupador sin representacion HTML. Solo sirve para agrupar elementos (el div agrupa elementos y tiene representacion HTML).-->
      <details>
        <summary>
          <router-link-stub to="/project/3"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/3"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/3"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
    <li>
      <!--template aquí sirve como agrupador sin representacion HTML. Solo sirve para agrupar elementos (el div agrupa elementos y tiene representacion HTML).-->
      <details>
        <summary>
          <router-link-stub to="/project/4"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
  </ul>
</aside>"
`;
